### PR TITLE
Add support for further granularity in #UNIQUE

### LINKED ISSUE #711: #UNIQUE by minute, hour, or day
The #UNIQUE will allow unique results by a set of fields.  However some fields (such as a datetime field) could use further granularity within the #UNIQUE function.  For example one may want a unique result for every hour of the day (e.g. #UNIQUE(HOUR(datetime_field)) or something like that.

### PR DESCRIPTION
**Pull request by lbschanno**
Begin to add support for additional granularity levels for the `#unique` 
function such that a user may get a unique result for every day, hour, or minute.

The following queries will be supported:
Query | Purpose
-------- | -----------
`#unique(field)` | Find unique results for every value of `field`.
`#unique(field[ALL])` | Identical to above.
`#unique(datetimeField[DAY])` | Find every unique result of `datetimeField` to the day.
`#unique_by_day(datetimefield)` | Identical to above.
`#unique(datetimeField[HOUR])` | Find every unique result of `datetimeField` to the hour.
`#unique_by_hour(datetimeField)` | Identical to above. 
`#unique(datetimeField[MINUTE])` | Find every unique result of `datetimeField` to the minute.
`#unique_by_minute(datetimeField)` | Identical to above.
`#unique(field,datetimeField1[DAY],datetimeField2[HOUR],datetimeField3[MINUTE])` | Compound queries are supported.
`#unique(field,datetimeField[DAY,HOUR])` | Consecutive granularities for the same field are supported.

Some additional notes:
* Query parsing is case-insensitive.
* If a value cannot be parsed as a date for a DAY, HOUR, or MINUTE granularity level, the original value will be kept and returned as a unique result.
* I tried to make it relatively straightforward to define additional granularity levels in the future if desired.
* Multiple granularity levels can be specified for the same field in different ways. The following queries are equivalent:
   * `#unique(field[HOUR,MINUTE])`
   * `#unique(field[HOUR],field[MINUTE])`
   * `#unique(field[HOUR]) AND #unique(field[MINUTE])`
   * `#unique(field[HOUR]) AND #unique_by_minute(field)`
   * `#unique_by_hour(field) AND #unique_by_minute(field)`

Fixes #711

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java [hunk 1]
```diff
@@ -45,19 +46,28 @@
...
-    private Set<String> fields;
+    private final BloomFilter<byte[]> bloom;
+    private UniqueFields uniqueFields;
     private Multimap<String,String> modelMapping;
     
-    public UniqueTransform(Set<String> fields) {
-        this.fields = deconstruct(fields);
+    public UniqueTransform(UniqueFields uniqueFields) {
+        this.uniqueFields = uniqueFields;
+        uniqueFields.deconstructIdentifierFields();
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java:55]: make this explicit ==> `this.uniqueFields.deconstructIdentifierFields();`

File: warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java [hunk 2]
```diff
@@ -286,105 +307,60 @@ public int compareTo(FieldSet o) {
...
-     * Get the grouping context for a field
-     * 
-     * @param documentField
-     * @return the grouping context (first and last elements of the group). If no grouping context, then a unique group is given for this fieldname.
-     */
-    private String getGroupingContext(String documentField) {
-        String[] parts = StringUtils.split(documentField, '.');
+    // Return the grouping context for the provided field if it exists. If no grouping context is returned, then field.ungrouped is returned.
+    private String getGroupingContext(String field) {
+        String[] parts = StringUtils.split(field, '.');
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java:331]: Consider swapping the datawave string utils for apache commons. Feel free to leave as-is.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import org.junit.After;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class QueryOptionsFromQueryVisitorTest {
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java:16]: Thank you for adding this test class

File: warehouse/query-core/src/test/java/datawave/query/language/parser/jexl/TestLuceneToJexlQueryParser.java [hunk 4]
```diff
@@ -10,6 +10,8 @@
 import org.junit.Before;
 import org.junit.Test;
 
+import static org.junit.Assert.assertEquals;
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/language/parser/jexl/TestLuceneToJexlQueryParser.java:13]: +1 for static test imports

**Review by apmoriarty — COMMENTED** [review]
Mostly cosmetic comments, overall form and function look good, tests look sound. I would love to see some of the documentation in the initial pull request description make it's way into the code.

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+     */
+    public void putAll(Multimap<String,UniqueGranularity> fieldMap) {
+        if (fieldMap != null) {
+            this.fieldMap.putAll(fieldMap);
+        }
+    }
+    
+    /**
+     * Return a copy of the fields contained within this {@link UniqueFields}. Modifications to this set will not modify the fields in this {@link UniqueFields}
+     * .
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+import org.apache.commons.lang.StringUtils;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+
+public class UniqueFields implements Serializable {
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java:26]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import datawave.data.normalizer.DateNormalizer;
+import org.apache.log4j.Logger;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.function.Function;
+
+public enum UniqueGranularity {
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java:16]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByHour.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+import datawave.query.jexl.functions.QueryFunctions;
+import datawave.query.language.functions.QueryFunction;
+import datawave.webservice.query.exception.BadRequestQueryException;
+import datawave.webservice.query.exception.DatawaveErrorCode;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+
+public class UniqueByHour extends JexlQueryFunction {
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByHour.java:14]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByMinute.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+import datawave.query.jexl.functions.QueryFunctions;
+import datawave.query.language.functions.QueryFunction;
+import datawave.webservice.query.exception.BadRequestQueryException;
+import datawave.webservice.query.exception.DatawaveErrorCode;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+
+public class UniqueByMinute extends JexlQueryFunction {
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByMinute.java:14]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+import datawave.query.jexl.functions.QueryFunctions;
+import datawave.query.language.functions.QueryFunction;
+import datawave.webservice.query.exception.BadRequestQueryException;
+import datawave.webservice.query.exception.DatawaveErrorCode;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+
+public class UniqueByDay extends JexlQueryFunction {
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java:14]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java (hunk 1 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java]: Fixed.

File: warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java (hunk 2 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java:305]: Sure, I don't have a particular attachment either way. Swapped for apache commons.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java (hunk 3 above)
Comment (lbschanno) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java:16]: No problem!

**Review by apmoriarty — APPROVED** [review]
LGTM

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+            return new UniqueFields();
+        }
+        
+        UniqueFields uniqueFields = new UniqueFields();
+        
+        final int finalIndex = string.length() - 1;
+        int currentIndex = 0;
+        String field = null;
+        
+        while (currentIndex < finalIndex) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java:63]: I feel this code could be quite a bit simpler.  Why not simply split on commas (StringUtils.split(string, ',')) and then parse each field separately.  This code I think will work but is a little difficult to maintain.

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    TRUNCATE_TEMPORAL_TO_HOUR("HOUR", new DateTimeValueFormatter("yyyy-MM-dd'T'HH")),
+    
+    /**
+     * A {@link UniqueGranularity} implementation that, if provided a datetime value, will return the datetime truncated to the minute. Otherwise, the original
+     * value will be returned.
+     */
+    TRUNCATE_TEMPORAL_TO_MINUTE("MINUTE", new DateTimeValueFormatter("yyyy-MM-dd'T'HH:mm"));
+    
+    private final String name;
+    private final Function<String,String> function;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java:42]: I love this.  This makes it easy to add additional enumerations down the road.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        super(QueryFunctions.UNIQUE_BY_DAY_FUNCTION, new ArrayList<>());
+    }
+    
+    /**
+     * query options contain a list of fields. Cannot be the empty list.
+     * 
+     * @throws IllegalArgumentException
+     */
+    @Override
+    public void validate() throws IllegalArgumentException {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java:26]: Can this function be moved a super class (e.g. UniqueByFunction)?  It appears the same in the other UniqueBy... classes.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        if (this.parameterList.isEmpty()) {
+            BadRequestQueryException qe = new BadRequestQueryException(DatawaveErrorCode.INVALID_FUNCTION_ARGUMENTS, MessageFormat.format("{0}", this.name));
+            throw new IllegalArgumentException(qe);
+        }
+    }
+    
+    @Override
+    public String toString() {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java:34]: Can this function be moved a super class (e.g. UniqueByFunction)? It appears the same in the other UniqueBy... classes.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java (hunk 12 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java:42]: Awesome, glad you like it.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java (hunk 13 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java]: Done.

File: warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java (hunk 14 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/language/functions/jexl/UniqueByDay.java]: Done.

File: warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java (hunk 11 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java:61]: I've simplified it to some degree, but I can't rely entirely on `StringUtils.split(string,',')` due to some edge cases. Let me know if you think it needs additional work.

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java:61]: ah right, you could have a comma inside the brackets.... I see now

**Review by ivakegg — DISMISSED** [review]

**Review by ivakegg — APPROVED** [review]

**Review by jwomeara — APPROVED** [review]

**Pull request merged**